lua/lint.lua

```diff
@@ -103,16 +103,39 @@ function M.lint(linter, client_id)
     cwd = vim.fn.getcwd(),
     detached = true
   }
-  assert(linter.cmd, 'Linter definition must have a `cmd` set: ' .. vim.inspect(linter))
-  handle, pid_or_err = uv.spawn(linter.cmd, opts, function(code)
+  local cmd = linter.cmd
+  if type(cmd) == 'function' then
+    cmd = cmd(bufnr)
+  end
+  -- When the function returns a table, the first value becomes the new
+  -- command. All other values are added to the start of the arguments list.
+  -- This makes it possible to dynamically switch between using for example
+  -- `rubocop` and `bundle exec rubocop` as the command.
+  if type(cmd) == 'table' then
+    local new_args = {}
+    local new_cmd = table.remove(cmd, 1)
+
+    for _, arg in ipairs(cmd) do
+      table.insert(new_args, arg)
+    end
+
+    for _, arg in ipairs(opts.args) do
+      table.insert(new_args, arg)
+    end
+
+    cmd = new_cmd
+    opts.args = new_args
+  end
+  assert(cmd, 'Linter definition must have a `cmd` set: ' .. vim.inspect(linter))
+  handle, pid_or_err = uv.spawn(cmd, opts, function(code)
```

Comment:
How about returning a tuple from the cmd function? The first return value is always a string, the second return value is optional and a list

`cmd, args = cmd()` 

Would simplfy this quite a bit.

What I'm unsure is if the `args` here are then an override or an extension of whatever is in the `linter.args` property.

Reply:
@mfussenegger Ah, I forgot Lua supports multiple return values. 

The goal of the arguments here was to act as an extension of `linter.args`. But the more I think of it, the more I think it makes more sense to unify both into a single function (as I suggested in https://github.com/mfussenegger/nvim-lint/pull/20#issuecomment-841384109). What do you think of that approach?

Reply:
It sounds like the `gitlint` case could be solved by supporting `args` as a function? In both scenarios the `cmd` is `gitlint`.

I'm not to keen on conflating both into the `args` just to then split it apart again.

Reply:
@mfussenegger I don't think that approach would work too well. For `gitlint` it would, because the base command is always `gitlint`. But for RuboCop (and possibly other linters) it means we have to repeat the same logic (a check to see if `bundle exec rubocop` is needed instead of just `rubocop`) in the `cmd` and `args` function. So something like this:

```lua
cmd = function()
  if use_bundler()
    return 'bundle'
  else
    return 'rubocop'
  end
end,
args = function(path)
  if use_bundler()
    return { 'exec', 'rubocop', path, ... }
  else
    return { path, ... }
  end
end
```

Unifying the two somehow means only a single check is necessary, and no repeated/redundant work needs to be performed.

Reply:
What would you think of supporting having the whole linter being a function? Instead of returning the linter table object, a linter can return a function that if called returns the linter table. So all options can be dynamically created if desired.